### PR TITLE
Initial Implementation of Tasks in CMS Configurations

### PR DESCRIPTION
**Pull request by wddgit**
This includes the initial implementation of the Task feature in the Core CMS configuration code. It allows creating Tasks, adding modules to them, and associating them to sequences, paths, end paths, and the schedule. In many ways Tasks behave like Sequences except that the order modules appear on Tasks does not matter. New unit tests were added to test the Core code changes.

This PR also includes changes to migrate a lot of configuration code to use Tasks. Large parts of PAT configuration code, all PAT tools, and some parts of RECO code were migrated. In particular, code that was used by runTheMatrix.py tests, IB unit tests, and other test configurations and that was already using unscheduled execution mode was migrated. Although large numbers of configuration files were migrated, an even larger number of configuration files were not migrated yet. There remains much migration work to be done. But the unmigrated configurations are not broken by this PR and should continue to run successfully. At a future date, the other code can be incrementally migrated as resources become available to do the work.

This PR modifies the convertToUnscheduled function to work properly with the new Task feature. EDProducers and EDFilters removed from paths and endpaths are automatically placed on a Task.

There are also changes to ConfigBuilder.py.

This is the third PR submitted for this set of changes. The first 2 were closed to make significant revisions. The first  one was PR #16538 and there are many comments from Chris in that PR which have all been addressed in separate commits after the first 3 commits. The second one was PR #16925, which contains Slava's comments and a lot of discussion that followed.

### CODE REVIEW

**Comment by wddgit** [comment]
The following changes have been made since I closed the previous PR:

1. Made changes requested by Slava and discussed at the Core meeting a few weeks ago. Where practical, the Task is now built from a list of modules and then a Sequence is created that does nothing more than associate the Task to the Sequence.

2. Previously, test configurations broken by this PR that were in unit tests had been fixed. Now I've also fixed all the test configurations that are not in unit tests but run without a fatal exception or seg fault. Of the test configurations that I found to be already broken, I fixed the simple ones and added a comment to the more complex cases. (You can identify a test configuration needing modification by looking for the allowUnscheduled parameter set True. The parameter itself isn't the problem but configurations with the parameter are ones that will need to add and associate tasks)

3. Fixed module ordering problems that showed up in the additional tests I did.

4. Fixed some instances where modules were imported only to be cloned so they would not be attached to the process.

5. Rebased as of an IB last Friday and fixed some issues and conflicts that showed up while merging in all those changes.

The Core parts of this PR have not changed at all since it was closed.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by Dr15Jones** [comment]
The unit test failures are not from this pull request.

File: HLTrigger/Configuration/python/CustomConfigs.py [hunk 1]
```diff
@@ -110,21 +110,13 @@ def HLTDropPrevious(process):
...
 
 def MassReplaceParameter(process,name="label",old="rawDataCollector",new="rawDataRepacker",verbose=False):
-#   replace values of named parameters
-    from PhysicsTools.PatAlgos.tools.helpers import massSearchReplaceParam
-    for s in process.paths_().keys():
-        massSearchReplaceParam(getattr(process,s),name,old,new,verbose)
-    for s in process.endpaths_().keys():
-        massSearchReplaceParam(getattr(process,s),name,old,new,verbose)
+    from PhysicsTools.PatAlgos.tools.helpers import massReplaceParameter
+    massReplaceParameter(process, name, old, new, verbose)
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/CustomConfigs.py]: What is the reason for this change?
For some reason we looped explicitly over both paths and endpaths!

**Comment by wddgit** [comment]
Hi Martin,

First, I noticed that there were functions MassReplaceInputTag and MassReplaceParameter that looped over all modules in all Paths and EndPaths and changed either InputTags or some configuration parameter. It appeared that they were trying to modify all the modules that might run. After this PR is merged, there can be modules on a Task that is directly associated with the Schedule. Those modules can run and will not be on any Path or EndPath, so the previous loops would miss them. That is what motivated me to modify these functions. I added a loop over the Tasks associated with the Schedule to the loops over Paths and EndPaths.

Second, as I was making this change I noticed that the function MassReplaceInputTag was duplicated in two places:

https://cmssdt.cern.ch/lxr/source/Configuration/Applications/python/ConfigBuilder.py#0173
and
https://cmssdt.cern.ch/lxr/source/HLTrigger/Configuration/python/CustomConfigs.py#0112

And both already call a function from PhysicsTools/PatAlgos to do their job:

https://cmssdt.cern.ch/lxr/source/HLTrigger/Configuration/python/CustomConfigs.py#0114

It seemed logical to me to refactor and eliminate the duplication and move the function with the loops to PhysicsTools/PatAlgos. And since I did this with MassReplaceInputTag, I decided to do the same thing for MassReplaceParameter. Note that the loop over EndPaths and Paths still exists, it was just moved into a new function in PhysicsTools/PatAlgos to eliminate duplication.

If you do not like this I can do it a different way. Let me know what you want here if you want me to change it.

**Comment by fwyzard** [comment]
Hi David,
only comment I have is that to me it would make more sense to move the function from PhysicsTools/PatAlgos to FWCore/ParameterSet .

**Comment by wddgit** [comment]
I think moving it to FWCore/ParameterSet is a good idea. I will also have to move the other function it calls and the visitor class that uses. And at the same time there is another almost identical function that also probably also ought to be moved.

https://cmssdt.cern.ch/lxr/source/RecoHI/HiEgammaAlgos/python/HiHelperTools.py#0213

This pull request is enormous already. Unless someone objects, I would like to do this as a separate pull request after this one gets merged.

**Comment by fwyzard** [comment]
> This pull request is enormous already. Unless someone objects, I would
like to do this as a separate pull request after this one gets merged.

Yes, of course.

**Comment by wddgit** [comment]
The changes in this PR are now integrated into the DEVEL IB, starting with CMSSW_9_0_DEVEL_X_2017-02-08-1700. I cannot see any problems there related to the changes in this PR.

**Comment by davidlange6** [comment]
hi @wddgit - not surprisingly this PR needs a rebase..

**Comment by wddgit** [comment]
I am working on the rebase. Yesterday I made the code changes affecting 4 files, but was waiting for a new IB to test with. I got sidetracked on some unrelated things today. I will probably push the fixes for the conflicting development Monday.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by wddgit** [comment]
The unit test failures are unrelated to this PR.

**Comment by Dr15Jones** [comment]
+1

**Comment by Dr15Jones** [comment]
2017-02-14 ORP decided this pull request should not go into CMSSW_9_0 but instead wait until CMSSW_9_1_X is open.

**Comment by slava77** [comment]
it looks like this PR will go to 91X.
So, it's somewhat on hold.

**Comment by wddgit** [comment]
I will continue to rebase this periodically and resolve conflicts until 91X is open. If anyone has any comments or issues, it would be nice to get them now so I could resolve them before 91X is open and this could get merged quickly into 91X when it opens. I am not aware of any unresolved issues at this point.

**Comment by slava77** [comment]
@wddgit 
just in case you haven't noticed already, there is a conflict in  Configuration/StandardSequences/python/Validation_cff.py

**Comment by davidlange6** [comment]
hold

for 91x. (but please add any concerns in the next week, this PR is in the DEVEL builds already)

**Comment by wddgit** [comment]
I just force pushed an update that resolved the conflict and rebased the commits in the PR on the same version as in IB 2-20-2017.

**Comment by Dr15Jones** [comment]
please test

**Comment by wddgit** [comment]
The one unit test failure is not related to this PR.

**Comment by wddgit** [comment]
Rebased to same commit as CMSSW_9_0_X_2017-02-26-2300

**Comment by Dr15Jones** [comment]
@davidlange6 with pre5 out, do you have an estimate on when the `master` branch will be created?

**Comment by slava77** [comment]
assign upgrade
just because this touches essentially all areas

**Comment by davidlange6** [comment]
for 91x
